### PR TITLE
Introduce new MemoryRecording for use with Jupyter notebooks

### PR DESCRIPTION
**Pull request by jleibs**
Builds on top of:
 - https://github.com/rerun-io/rerun/pull/1798

Rather than implicitly draining the contents of the BufferSink this introduces a new MemorySink that uses a separate (Arc'd storage). This storage is wrapped with a python class that is now responsible for retrieving the in-memory rrd.

The MemoryRecording now implements `_repr_html_` and so can be displayed using either `rec` or `rec.show()`

image

All `rr` calls still log the the last-created memory_recording (the default sink). In the future, we will allow the sink to establish context for rr log calls, or allow the user to more explicitly control which sink is the default.

The main benefit is this allows the user to hang onto one or more recording objects and be explicit about the context in which they are shown. This will also facilitate combining the same rrd object with different blueprints.

Even now this means that a recorded option can be shown multiple times with different arguments such as canvas height and width, whereas the previous inline_show() operation consumed the data requiring you to re-log it again to set up a different canvas.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)

### CODE REVIEW

File: rerun_py/src/python_session.rs [hunk 1]
```diff
@@ -200,17 +200,20 @@ impl PythonSession {
...
-        let backlog = self.sink.drain_backlog();
-        // Since all the messages have been drained we need to send the recording-msg again.
-        self.has_sent_begin_recording_msg = false;
+    /// Send all pending and future log messages to an in-memory store
+    pub fn memory_recording(&mut self) -> rerun::sink::MemorySinkStorage {
+        let memory_sink = rerun::sink::MemorySink::default();
+        let buffer = memory_sink.buffer();
 
-        let mut buffer = Cursor::new(Vec::new());
+        if !self.enabled {
```
Comment (emilk) [inline rerun_py/src/python_session.rs:208]: do this before setting up the sink?

File: crates/re_sdk/src/log_sink.rs [hunk 2]
```diff
@@ -76,6 +76,50 @@ impl LogSink for BufferedSink {
...
 }
 
+/// The storage used by [`MemorySink`]
+#[derive(Default, Clone)]
+pub struct MemorySinkStorage(std::sync::Arc<parking_lot::Mutex<Vec<LogMsg>>>);
+
+///
+impl MemorySinkStorage {
+    /// Lock the contained buffer
+    pub fn lock(&self) -> parking_lot::MutexGuard<'_, Vec<LogMsg>> {
```
Comment (emilk) [inline crates/re_sdk/src/log_sink.rs:86]: I don't see this being used outside this function - no need to leak internals imho
```suggestion
   fn lock(&self) -> parking_lot::MutexGuard<'_, Vec<LogMsg>> {
```

File: crates/re_sdk/src/log_sink.rs [hunk 3]
```diff
@@ -76,6 +76,50 @@ impl LogSink for BufferedSink {
     }
 }
 
+/// The storage used by [`MemorySink`]
```
Comment (emilk) [inline crates/re_sdk/src/log_sink.rs:79]: This comment would make more sense if you declared `MemorySinkStorage` _after_ `MemorySink` 😉

File: crates/re_sdk/src/log_sink.rs [hunk 4]
```diff
@@ -76,6 +76,50 @@ impl LogSink for BufferedSink {
...
+        let mut buffer = std::io::Cursor::new(Vec::new());
+        re_log_encoding::encoder::encode(messages.iter(), &mut buffer)?;
+        Ok(buffer.into_inner())
+    }
+}
+
+/// Store log messages directly in memory
+///
+/// Unlike `BufferedSink` this uses an external buffer that can be
+/// accessed directly.
```
Comment (emilk) [inline crates/re_sdk/src/log_sink.rs:102]: Can we replace `BufferedSink` with `MemorySink`?

**Review by emilk — APPROVED** [review]

File: crates/re_sdk/src/log_sink.rs (hunk 2 above)
Comment (jleibs) [reply crates/re_sdk/src/log_sink.rs:86]: Ahh good call -- went away once I moved `rrd_as_bytes` into this object.

**Pull request merged**